gh-96121: Merge sqlite3.Row examples (#96122)
python/cpython: Doc/library/sqlite3.rst

         >>> cur.connection == con
         True

.. The sqlite3.Row example used to be a how-to. It has now been incorporated
   into the Row reference. We keep the anchor here in order not to break
   existing links.

.. _sqlite3-columns-by-name:

.. _sqlite3-row-objects:

Row objects
^^^^^^^^^^^

.. class:: Row

   A :class:`!Row` instance serves as a highly optimized
   :attr:`~Connection.row_factory` for :class:`Connection` objects.

>    It tries to mimic a :class:`tuple` in most of its features,

Reviewer: Isn't it more like a `namedtuple` since it supports access by name and position?
```suggestion
   It tries to mimic a :class:`~collections.namedtuple` in most of its features,
```

Author: I'm afraid that mentioning named tuples will create assumptions that columns can be accessed using the dot operator; which is not true. See also #57508.

Author: I wonder if we should just drop that "it mimics" sentence, and just describe what it does (it can quack, walk, swim, instead of it mimics a duck)

Reviewer: > will create assumptions that columns can be accessed using the dot operator

Good point, I mixed up item access and attribute access.

> just describe what it does

This sounds good to me.

         >>> cur.connection == con
         True

.. The sqlite3.Row example used to be a how-to. It has now been incorporated
   into the Row reference. We keep the anchor here in order not to break
   existing links.

> .. _sqlite3-columns-by-name:
> 
> .. _sqlite3-row-objects:

Reviewer 2: ```suggestion
.. _sqlite3-columns-by-name:
.. _sqlite3-row-objects:
```

Trivial, but any reason to leave a space here? I've usually seen multiple ref target labels without one, IIRC.

   .. versionchanged:: 3.5
      Added support of slicing.

   Example::

      >>> con = sqlite3.connect(":memory:")
      >>> con.row_factory = sqlite3.Row

>       >>> res = con.execute("select 'Earth' as name, 6378 as radius")

Reviewer 2: ```suggestion
      >>> res = con.execute("SELECT 'Earth' AS name, 6378 AS radius")
```

Wouldn't this follow the SQL "style guide" you established in #95271 ?

Author: Yes, thanks! (I'm used to writing SQL all lower case, but for the docs, I think a best practices style is better.)

      >>> row = res.fetchone()
      >>> row.keys()
      ['name', 'radius']
      >>> row[0], row["name"]  # Access by index and name.
      ('Earth', 'Earth')

>       >>> row["RADIUS"]

Reviewer: ```suggestion
      >>> row["RADIUS"]  # Column names are case-insensitive.
```
I think this suggestion fell through the cracks (unless you think the comment is unnecessary).

Author: We already said that in the text, so repeating it might not be needed. OTOH, we repeat access by name and index in a comment in the example.

   .. versionchanged:: 3.5
      Added support of slicing.

Let's assume we initialize a table as in the example given above::

   con = sqlite3.connect(":memory:")
   cur = con.cursor()
   cur.execute('''create table stocks
   (date text, trans text, symbol text,
    qty real, price real)''')
   cur.execute("""insert into stocks
               values ('2006-01-05','BUY','RHAT',100,35.14)""")
   con.commit()

>    cur.close()

Reviewer: This was useful in case I wanted to play around in the interpreter.  Maybe you could have a sample db that you initialize one and then reuse throughout the docs?  It could be the Monty Python movies one from the tutorial, or perhaps another one defined in one of the howto or in a specific section before the reference.

Another solution is to use for the setup the collapsible block of code we discussed about, even though that's not supported by default by Sphinx.

Author: > Maybe you could have a sample db that you initialize one and then reuse throughout the docs?

I've though about that, yes. It will require a lot of changes, though.

Author: OTOH, it is pretty easy to set up a sample table, we might as well just do that.

>       >>> res = con.execute("select 'Tellus' as name, 6378 as radius")

Reviewer: ```suggestion
      >>> res = con.execute("select 'Tellus' as name, 6378 as radius")
```
I find this line a bit confusing.  It seems that instead of selecting an actual row from a db it just return the values you provide, even though it requires a connection (to what?).  In addition, people might not be familiar with `'Tellus'`, adding to the confusion.

Author: Yes, it uses a SELECT query with literals. The connection is needed to execute the query.

We can use another example; I just don't like the current example.

Author: We can create a test database as well, but that will expand the example with, IMO, unneeded information. We're interested in the row factory only; using a SELECT query that operates with literals quickly gets us a row of data without the need for a database table.

Reviewer: > The connection is needed to execute the query.

Is this just a connection to `:memory:`?  Since I'm not querying an existing table, I assume that I don't need to create it and populate it -- how many lines would it take to create a valid `con`?

The goal for me would be having a reproducible example that I can play around with in the interactive interpreter.

Are there other factories?  Maybe the example could be moved back in an howto that sets up a single db and compares different factories if you think an howto specific to row factories is overkill?

Author: > Is this just a connection to :memory:? 

With the type of query used here, that does not matter. It could be any kind of connection, with or without tables. Since we're SELECTing literals, it does not matter.

A minimal connection is:

```python3
import sqlite3
cx = sqlite3.connect(":memory:")
```

Author: > Are there other factories? Maybe the example could be moved back in an howto that sets up a single db and compares different factories if you think an howto specific to row factories is overkill?

Not at the moment, but there has been talk about a namedtuple-factory.

I think a general "how to use/write row factories" could be useful. There is a current example in the `row_factory` reference; we could build on that. (For another PR/issue, of course.)

Reviewer: > A minimal connection is:

Since the `import sqlite3` can be omitted, this means that a single line (`con = sqlite3.connect(":memory:")`) is enough to turn that into a working example, right?
If that's the case, having a working example outweighs the extra line IMHO.

The `SELECT` with literals is still a bit "weird" if someone is not familiar with it, but the full example will help showing that it's performed on an empty DB and is not assuming the existence of an existing table.  If you want to be more explicit, perhaps a footnote could be added.

Author: Yes, the SELECT trick may feel weird if you are unfamiliar with it. It may make sense to add a comment or footnote.

Reviewer 2: A comment might probably be most appropriate, IMO, since it explains a potentially confusing but specific implementation detail of the example

Author: Hm, the SELECT-with-literals trick is very convenient for a lot of examples. Perhaps it would be better to add a "tips and tricks" howto, where we show it, and just reference that howto from the examples that use it?[^1] Obviously, it would not make sense to explain it every time we use it, hence my slight reluctance to add a note about it in this particular example.

[^1]: it could be a footnote too

Author: I'm not sure how to best resolve this _right now_. The old example also used SELECT-with-literals, so IMO, we're not introducing more confusion than there possibly was before. We can always improve matters with future PRs. I'll create a ticket for this on the tracker, though, so we don't forget about it.

Author: I created #96165

>       >>> row = res.fetchone()

Reviewer: Here you could show the `repr` after fetching the line, since it's mentioned in the docs above and because it confirms that you got a `Row` back.

Author: Good point.

Author: Ah, I remember; it does _not_ support `repr`. There was a long discussion regarding `repr`, though. It was concluded that it did not make sense to add support for it. See #80070

>       'Tellus'

Reviewer: If you want to make it a bit more compact:
```suggestion
      >>> row[0], row["name"]  # supports access by both index and name
      ('Tellus', 'Tellus')
```
I also like to add inline comments that tells what's the point of the example (unless it's obvious).

Author: Yes, that's neat.

      >>> row["RADIUS"]

>       6378

Reviewer: ```suggestion
      >>> row["RADIUS"]  # field names are case-insensitive
      6378
```
(Is `fields` the right term here?)

Author: We should use column name for consistency.

      >>> "Tellus" in row

>       True

Reviewer: Is this a useful operation?
Knowing that the row is `tuple`-like, I would expect this to work even without an example, but on the other hand I can't think of a common situation where I would need to check for this.

Author: I'm not sure if it is really needed. We could probably get rid of it.
